Rubberduck.VBEditor.VB6/SafeComWrappers/Office/CommandBarButton.cs

```diff
@@ -0,0 +1,343 @@
+extern alias Office_v8;
+using System;
+using System.Drawing;
+using System.Runtime.InteropServices;
+using System.Windows.Forms;
+using Microsoft.CSharp.RuntimeBinder;
+using NLog;
+using Rubberduck.VBEditor.SafeComWrappers.Abstract;
+using MSO = Office_v8::Office;
+using VB = Microsoft.Vbe.Interop.VB6;
+
+// ReSharper disable once CheckNamespace - Special dispensation due to conflicting file vs namespace priorities
+namespace Rubberduck.VBEditor.SafeComWrappers.Office8
+{
+    public class CommandBarButton : SafeRedirectedEventedComWrapper<MSO.CommandBarButton, VB.CommandBarEvents, VB._dispCommandBarControlEvents>, ICommandBarButton, VB._dispCommandBarControlEvents
+    {
+        private readonly CommandBarControl _control;
+        private readonly IVBE _vbe;
+        private static readonly Logger _logger = LogManager.GetCurrentClassLogger();
+        public const bool AddCommandBarControlsTemporarily = false;        
+
+        // Command bar click event is sourced from VBE.Events.CommandBarEvents[index]
+        // where index is the command bar button COM object.
+        public CommandBarButton(MSO.CommandBarButton target, IVBE vbe, bool rewrapping = false) 
+            : base(target, ((VB.VBE)vbe.HardReference).Events.CommandBarEvents[target], rewrapping)
```

Comment:
This will certainly leak RCWs. You cannot chain member accesses on COM interop objects without leaking all intermediate objects. This is why we wrap the result of each member call in the COM wrappers.

Reply:
Raised #3949